### ISSUE TITLE
Define policy on error handling

### ISSUE THREAD

**Issue by sigiesec** [zeromq/libzmq#3148]
Currently, the policy on using assertions and errors is not documented and somewhat unclear.

There are three kinds of "error handling":
1. plain assert (at least for cmake, these are active in Debug builds only)
2. zmq_assert, errno_assert, wsa_assert, alloc_assert (these are active in both Debug and Release)
3. errors signalled to the user via errno/zmq_errno()
4. errors signalled via socket monitor events

Options 1+2 vs. 3 vs. 4 are used consistently. More or less. Maybe this is hard to determine without a clear policy.

However, option 1 vs. 2 is really quite unclear, and I think this should be improved, with the intent of moving more towards option 1. The problem with option 2 is that it bloats the production code, in terms of both code *size* (quite significantly as each occurrence produces some string constants) and *efficiency* (by preventing optimizations, in particular inlining). As a general idea, option 2 makes sense to handle unexpected responses from OS/library functions, but it should not be used to check if internal libzmq calls are called correctly, e.g. This should better be done with plain asserts IMO.

The policy should be documented, since it is not a universal standard.

**Comment by bluca** [comment]
> plain assert (these are active in Debug builds only)

Note that this is true only for Windows

**Comment by sigiesec** [comment]
Hm?

That assert(...) evaluates to (void)0 in release builds (i.e. defining NDEBUG) is specified by the standard. See, e.g. http://en.cppreference.com/w/cpp/error/assert (ok, this isn't the standard, but I am pretty sure that is the source of it). And at least the Linux/clang analysis build was a Release (in the sense of CMAKE_BUILD_TYPE) build before, and it complained about several unused stored values, which were only referenced in assert calls. These are gone since I changed to CMAKE_BUILD_TYPE to Debug.

**Comment by bluca** [comment]
Yes, but NDEBUG is never defined by any distro AFAIK (note that I'm talking about autotools here)

**Comment by bluca** [comment]
See for example the build log on Debian: https://buildd.debian.org/status/fetch.php?pkg=zeromq3&arch=i386&ver=4.2.5-1&stamp=1522196748&raw=0

**Comment by sigiesec** [comment]
Oh, really? That is really surprising for me. Is this done on purpose?

As I wrote in my original comment, this leads to inefficient code, as a function containing an active assert will most likely never be inlined, and other optimizations may also be prevented. This probably also applies to STL code that is used.

However, the cmake builds do define NDEBUG on any platform when doing a Release build, which is the default. See, e.g., here: https://travis-ci.org/sigiesec/libzmq/jobs/386094823#L564

**Comment by bluca** [comment]
Distributions leave it to the upstreams what to use, as it can break things unexpectedly, and generally few flags are pre-defined, and it's usually -O2 -g and various security-related hardening flags (eg: pie/pic to enable address layout randomisation, and even those caused huge headaches when they were turned on by default).
We shouldn't define it either, as in the grand scheme of things it's much more important to clearly guard against illegal API usage or throw an error when an unrecoverable situation arises, even if the price is a little less efficient generated code.

Users that need to really squeeze every last clock cycle can rebuild it themselves, and will most likely use more aggressive compiler optimisations than the defaults anyway.

**Comment by sigiesec** [comment]
I saw there was some debate on cmake defining NDEBUG on the Debian mailing list some time ago when cmake changed (harmonized) its behaviour with version 2.8.11. And they clearly say that Debian policy is expressly NOT to define NDEBUG.

Well... I am still somewhat surprised. In my experience, not defining NDEBUG has more significant downside effects on performance than can be described by "squeere every last clock cycle", but I have no concrete data right now :)

I updated my original post, however.

Apart from what is typically done, I still think there should be an option to build with *disabled plain asserts* in the sense of no. 1 but *enabled custom *_assert* in the sense of no. 2.
